### PR TITLE
win,fs: fix inconsistencies between platforms for rename

### PR DESCRIPTION
**Pull request by Shivang44**
Renaming directories has odd inconsistencies between Unix and Win
platforms. For example, renaming dirA to dirB when dirB is empty fails
on Win while succeeds on Linux. This commit fixes that by removing dirB
if dirB is empty. Additionally, renaming dirA to dirB when dirB is NOT
empty, on Linux, returns a ENOTEMPTY error, while on Win, returns a
EPERM error. This fixes this consistency be returning a ENOTEMPTY error
on windows as well.

Related to nodejs/node#21957

Note to reviewers: I am very new to C/C++ and I am sure I made mistakes,
both in my method and my test case. I would appreciate any in-depth feedback
on my code. Thank you!

Note about tests: The tests pass, but on some rare occasions they will fail. I assuming there are just flaky tests?

### CODE REVIEW

**Comment by Shivang44** [comment]
I realize windows is not that popular of platform for nodejs, but I feel fixing these inconsistencies is important. Can somebody please review this PR?

File: src/win/fs.c [hunk 1]
```diff
@@ -1301,6 +1304,34 @@ static void fs__fstat(uv_fs_t* req) {
 
 
 static void fs__rename(uv_fs_t* req) {
+  WCHAR* pathw = req->file.pathw;
```
Comment (cjihrig) [inline src/win/fs.c]: This variable appears to be unused.

**Comment by Shivang44** [comment]
@cjihrig Thanks for reviewing the PR! I will revise the PR according to your comments, but want to discuss it before I do so time is not wasted if this change is deemed to be not right by other maintainers. 

> Overall, I'm a little torn on this. If it aligns the behavior across platforms, that's good. However, it also introduces a bit of a potential race condition due to the extra file system calls. There is also the question of whether this would classify as a bug fix or breaking change. I'd like to know what other collaborators think.

Which part of the code creates a race condition? It was my understanding that the calls like PathIsDirectory RemoveDirectory block until they are returned? So there wouldn't be a race condition right? Again, my understanding of these calls and C are limited.

> It depends which test(s) you're talking about. Can you be more specific about individual tests and platforms?

I will rerun the tests when I get a chance and let you know!

**Comment by refack** [comment]
Hello @Shivang44,

> I feel fixing these inconsistencies is important.

I'm sure you are aware there are numerous other inconsistencies, for example; case sensitive names, timestamp accuracy, no links on FAT, different ACL mechanisms.

> For example, renaming dirA to dirB when dirB is empty fails on Win while succeeds on Linux. This commit fixes that by removing dirB if dirB is empty.

IMHO implementing the complex solution you suggest should be left for downstream projects (e.g. `node` like [`mkdirp`](https://github.com/nodejs/node/pull/21875), or a userland module)

As for returning `ENOTEMPY` I think that is a breaking change. Since you are already making breaking changes, I suggest returning `ENOTEMPY` for both cases (if `dirB` is empty or not). I would go as far as to suggest to return `ENOTEMPY` even on POSIX, but this requires buy-in from the Collaborators.

**Comment by refack** [comment]
P.S. I just found what I believe is a bug on Windows:
```cmd
> mkdir dirA
> echo " " > dirB
> node -e "fs.renameSync('dirA', 'dirB')"

>
```
This will overwrite `dirB`.

**Comment by Shivang44** [comment]
Due to the lack of interest as well as the above mentioned (potential issues), I will be closing this PR. If other maintainers think that this change would be a good addition to libuv, I would be happy to reopen it work on it!

Thanks for your comments and help @cjihrig @refack

**Pull request closed without merging**